propagators/aws/xray/sampled_unsampled_span_test.go

```diff
@@ -0,0 +1,203 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import (
+	"context"
+	"testing"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var tracer = otel.Tracer("sample-app")
+
+func startAndEndSampledSpan() {
+	var span trace.Span
+	_, span = tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+
+	defer span.End()
+}
+
+func startAndEndNestedSampledSpan() {
+	var span trace.Span
+	ctx, span := tracer.Start(context.Background(), "Parent operation...")
+	defer span.End()
+
+	_, span = tracer.Start(ctx, "Sub operation...")
+	defer span.End()
+}
+
+func getCurrentSampledSpan() trace.Span {
+	var span trace.Span
+	ctx, span := tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+	defer span.End()
+
+	return trace.SpanFromContext(ctx)
+}
+
+func addAttributesToSampledSpan() {
+	var span trace.Span
+	_, span = tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+	defer span.End()
+
+	span.SetAttributes(attribute.Key("example attribute 1").String("value 1"))
+	span.SetAttributes(attribute.Key("example attribute 2").String("value 2"))
+}
+
+func startAndEndUnSampledSpan() {
+	var span trace.Span
+	_, span = tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+
+	defer span.End()
+}
+
+func startAndEndNestedUnSampledSpan() {
+	var span trace.Span
+	ctx, span := tracer.Start(context.Background(), "Parent operation...")
+	defer span.End()
+
+	_, span = tracer.Start(ctx, "Sub operation...")
+	defer span.End()
+}
+
+func getCurrentUnSampledSpan() trace.Span {
+	var span trace.Span
+	ctx, span := tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+	defer span.End()
+
+	return trace.SpanFromContext(ctx)
+}
+
+func addAttributesToUnSampledSpan() {
+	var span trace.Span
+	_, span = tracer.Start(
+		context.Background(),
+		"Example Trace",
+	)
+	defer span.End()
+
+	span.SetAttributes(attribute.Key("example attribute 1").String("value 1"))
+	span.SetAttributes(attribute.Key("example attribute 2").String("value 2"))
+}
+
+func initialization() func() {
+	idg := NewIDGenerator()
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithIDGenerator(idg),
+	)
+
+	curTp := otel.GetTracerProvider()
+	curProp := otel.GetTextMapPropagator()
+
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(Propagator{})
+
+	return func() {
+		otel.SetTracerProvider(curTp)
+		otel.SetTextMapPropagator(curProp)
```

Comment:
These are no-ops. The global set operation is only ever performed once.

https://github.com/open-telemetry/opentelemetry-go/blob/651409851900230cc2f8c62a5ec69a7ac9307847/internal/global/state.go#L74

https://github.com/open-telemetry/opentelemetry-go/blob/651409851900230cc2f8c62a5ec69a7ac9307847/internal/global/state.go#L50

It would make more sense to define `tracer` with `TracerProvider` in an `init` function and skip any dealing with the `TextMapPropagtor` (it doesn't look used in any of these benchmarks).

Reply:
Are you suggesting to add [`trace.Start`](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/812/files#diff-29f7631aa697ee4a691f655bda8c68861c3c2020bcb3ad9ee2a70abe65063f48R103) in `init` function ? As per @Aneurysm9's suggestion I have add restoring logic of propagators so had to add those in every benchmark.

Reply:
I'm suggesting you create a `TracerProvider` in an `init` function that you then use to set the value of a package level `tracer` variable with.

The restoring logic you added does nothing, the second call is a no-op. You should not use the global tracer here.